apps/emqx/src/emqx_trace/emqx_trace_json_formatter.erl

```diff
@@ -57,12 +57,46 @@ prepare_key_value(K, {Formatter, V}, PEncode) when is_function(Formatter, 1) ->
         _:_ ->
             {K, V}
     end;
-prepare_key_value(K, {ok, Status, Headers, Body}, PEncode) when
-    is_integer(Status), is_list(Headers), is_binary(Body)
+prepare_key_value(K, {ok, {Formatter, V}}, PEncode) when is_function(Formatter, 1) ->
+    %% Unwrap
+    prepare_key_value(K, {Formatter, V}, PEncode);
+prepare_key_value(host, {I1, I2, I3, I4}, _PEncode) when
+    is_integer(I1),
+    is_integer(I2),
+    is_integer(I3),
+    is_integer(I4)
 ->
-    %% This is unlikely anything else then info about a HTTP request so we make
-    %% it more structured
-    prepare_key_value(K, #{status => Status, headers => Headers, body => Body}, PEncode);
+    %% We assume this is an IP address
+    {host,
+        unicode:characters_to_binary([
+            integer_to_binary(I1),
+            <<".">>,
+            integer_to_binary(I2),
+            <<".">>,
+            integer_to_binary(I3),
+            <<".">>,
+            integer_to_binary(I4)
+        ])};
+prepare_key_value(K, {ok, StatusCode, Headers}, PEncode) when
+    is_integer(StatusCode), StatusCode >= 200, StatusCode < 300, is_list(Headers)
+->
+    prepare_key_value(K, {ok, StatusCode, Headers, <<"">>}, PEncode);
+prepare_key_value(K, {ok, StatusCode, Headers, Body}, PEncode) when
```

Comment:
this seems to be the wrong module to format http results?

Reply:
In a way yes. However there are a few connectors that return the result in the same format so my idea was that it would make the code less cluttered if these results got formatted implicitly. It might be still be best to be more explicit? I'm okay with both so I can make these into lacy format functions in the connectors if you think this is better?

Reply:
Just realized that we also need to handle the format functions in the other log formanters as well!

Reply:
I'm having a discussion with @thalesmg . Returning the formatter in the result from the connector is a quite ugly hack that could lead to problems. His suggestion is to add a connector callback for trace formatting the result. The formatter module can infer which callback to use from the Action ID. I think this sounds like a reasonable suggestion so I will try to implement this tomorrow.

Reply:
Implemented a solution in line with @thalesmg's suggestion here: [e0d7a87](https://github.com/emqx/emqx/pull/12958/commits/e0d7a876ec4c92b800ae5bd9ff12e9cd1bc2e9a9)